platform/posix/ota_pal/source/ota_pal_posix.c

```diff
@@ -0,0 +1,626 @@
+            LogError( ( "Failed to allocate buffer memory." ) );
+            mainErr = OtaPalOutOfMemory;
+        }
+    }
+    else
+    {
+        LogError( ( "File signature check failed at INIT." ) );
+        mainErr = OtaPalSignatureCheckFailed;
+    }
+
+    return mainErr;
+}
+
+/* Verify the signature of the specified file using OpenSSL. */
+static OtaPalStatus_t otaPal_CheckFileSignature( OtaFileContext_t * const C )
+{
+    OtaPalMainStatus_t mainErr = OtaPalSuccess;
+    EVP_PKEY * pPkey = NULL;
+    EVP_MD_CTX * pSigContext = NULL;
+
+    if( C != NULL )
+    {
+        /* Extract the signer cert from the file */
+        pPkey = Openssl_GetPkeyFromCertificate( C->pCertFilepath );
+
+        /* Create a new signature context for verification purpose */
+        pSigContext = EVP_MD_CTX_new();
+
+        if( ( pPkey != NULL ) && ( pSigContext != NULL ) )
+        {
+            /* Verify an ECDSA-SHA256 signature. */
+            mainErr = Openssl_DigestVerify( pSigContext, pPkey, C->pFile, C->pSignature );
+        }
+        else
+        {
+            if( pSigContext == NULL )
+            {
+                LogError( ( "File signature check failed at NEW sig context." ) );
+                mainErr = OtaPalSignatureCheckFailed;
+            }
+            else
+            {
+                LogError( ( "File signature check failed at EXTRACT pkey from signer certificate." ) );
+                mainErr = OtaPalBadSignerCert;
+            }
+        }
+
+        /* Free up objects */
+        EVP_MD_CTX_free( pSigContext );
+        EVP_PKEY_free( pPkey );
+    }
+    else
+    {
+        /* FIXME: Invalid error code for a NULL file context. */
+        LogError( ( "Failed to check file signature: Paramater check failed: "
+                    " Invalid OTA file context." ) );
+        /* Invalid OTA context or file pointer. */
+        mainErr = OtaPalNullFileContext;
+    }
+
+    return OTA_PAL_COMBINE_ERR( mainErr, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+OtaPalStatus_t otaPal_ResetDevice( OtaFileContext_t * const C )
+{
+    ( void ) C;
+
+    /* Return no error.  linux implementation does not reset device. */
+    return OTA_PAL_COMBINE_ERR( OtaPalSuccess, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+OtaPalStatus_t otaPal_ActivateNewImage( OtaFileContext_t * const C )
+{
+    ( void ) C;
+
+    /* Return no error. linux implementation simply does nothing on activate.
+    * To run the new firmware image, double click the newly downloaded exe */
+    return OTA_PAL_COMBINE_ERR( OtaPalSuccess, 0 );
+}
+
+/* Set the final state of the last transferred (final) OTA file (or bundle).
+ * On linux, the state of the OTA image is stored in PlaformImageState.txt. */
+OtaPalStatus_t otaPal_SetPlatformImageState( OtaFileContext_t * const C,
+                                             OtaImageState_t eState )
+{
+    OtaPalMainStatus_t mainErr = OtaPalSuccess;
```

Comment:
Initialize return values to failure – see previous comment for reason:

```c
OtaPalMainStatus_t mainErr = OtaPalBadImageState;
```

Then rework the logic on lines 515-531 as follows (_note that this is also simpler & easier to read – a nice side effect of initializing return status to failure value_) and delete lines 536 and 543.

```c
if( 1UL == fwrite( &eState, sizeof( OtaImageState_t ), 1, pPlatformImageState ) )
{
    if( 0 == fclose( pPlatformImageState ) )
    {
        mainErr = OtaPalSuccess;
    }
    else
    {
        LogError( ( "Unable to close image state file." ) );
        subErr = errno;
    }
}
else
{
    LogError( ( "Unable to write to image state file. error-- %d", errno ) );
    subErr = errno;
}
```